### PR TITLE
Add CompilationUnit Cache

### LINKED ISSUE #14214: Caching layer to Parser, for caching the Compilation units
**Description:**
For each parse current parser implementation parses the source and when there are considerably a larger amount of files, parsing phase takes significant time and language server compilation phases lag.
Following two tasks need to address
- [x] Caching the compilation unit
- [x] Caching the Syntax diagnostics

**Affected Versions:**
v0.990.3 at least

### PR DESCRIPTION
**Pull request by hasithaa**
Fixes #14214 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 1]
```diff
@@ -0,0 +1,1906 @@
...
+    @Override
+    public void visit(BLangFunction source) {
+
+        BLangFunction clone = new BLangFunction();
+        source.cloneRef = clone;
+
+        clone.attachedFunction = source.attachedFunction;
+        clone.objInitFunction = source.objInitFunction;
+        clone.interfaceFunction = source.interfaceFunction;
+        clone.anonForkName = source.anonForkName;
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java]: Can't we move this into `cloneBLangInvokableNode` as well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 2]
```diff
@@ -0,0 +1,1906 @@
...
+
+/**
+ * Node Visitor for cloning AST nodes.
+ *
+ * @since 1.1
+ */
+class NodeCloner extends BLangNodeVisitor {
+
+    private static final CompilerContext.Key<NodeCloner> NODE_CLONER_KEY =
+            new CompilerContext.Key<>();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java]: Can be merged?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 3]
```diff
@@ -0,0 +1,1906 @@
...
+
+    private NodeCloner() {
+
+    }
+
+    public static NodeCloner getInstance(CompilerContext context) {
+
+        NodeCloner cleaner = context.get(NODE_CLONER_KEY);
+        if (cleaner == null) {
+            cleaner = new NodeCloner();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java]: Shouldn't this be added to the context?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/ParserCache.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return cache;
+    }
+
+    BLangCompilationUnit get(PackageID packageID, String entryName, int hash, int length) {
+
+        if (!cacheEnabled) {
+            return null;
+        }
+        Map<String, BLangCompilationUnit> sourceEntryCache;
+        if ((sourceEntryCache = pkgCache.get(packageID)) == null || !pkgCache.containsKey(packageID)) {
```
Comment (grainier) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/ParserCache.java]: If `pkgCache` doesn't contain `packageID`, `sourceEntryCache` would anyway be null right? So isn't `!containsKey` is redundant?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java (hunk 1 above)
Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java]: Design is to keep attribute clone with relevant visit method. If attribute extends from the superclass,  those are moved to the "clone<SupperClasName>" method.  In this case, these attributes are coming from BLangFunction node, hence kept with BLangFunction node visit method.

**Pull request merged**